File: frontend/ui-kernel/src/kui/ui/kernel/component/Numbers.scala
package kui.ui.kernel.component

/** Whole numbers, rendered so that a column of them can be compared down rather than read across.
  *
  * Digits are grouped in threes with a thin space rather than a comma. A comma is a decimal separator in
  * about half the world, and "1,204" read as a decimal is three orders of magnitude wrong — which, on a lag
  * figure an operator is deciding whether to act on, is the whole decision. The thin space is what the
  * international standard uses for exactly this reason, and it survives a copy-paste into a spreadsheet as a
  * space rather than as a wrong separator.
  *
  * There is no rounding to "1.2M". A lag of 1 204 331 and a lag of 1 249 000 both round to the same three
  * characters and they are not the same problem.
  */
object Numbers {

  /** The separator between groups of three digits: U+2009 THIN SPACE. */
  val GroupSeparator: String = " "

  def grouped(value: Long): String = {
    val digits = math.abs(value).toString
    val sign = if value < 0 then "-" else ""

    // Grouped from the right, which is where the units digit is, so a number of any length lines up under
    // any other.
    val chunks =
      digits.reverse.grouped(3).map(_.reverse).toList.reverse

    sign + chunks.mkString(GroupSeparator)
  }

  /** A rate, at one decimal place, with an explicit sign when it is negative.
    *
    * One decimal, not three: this is a rate sampled over a thirty-second interval, and the digits past the
    * first are noise from where in the interval the two samples happened to land. Printing them would invite
    * an operator to compare two numbers that do not differ.
    *
    * A rate below one tenth of a record per second but not zero renders as `< 0.1` rather than as `0.0`,
    * because a group committing something very slowly and a group committing nothing are the difference
    * between "slow" and "stuck", which is the whole question this column is asked.
    */
  def rate(value: Double): String =
    if value == 0.0 then "0"
    else if math.abs(value) < 0.05 then if value < 0 then "> -0.1" else "< 0.1"
    else {
      val rounded = math.round(math.abs(value) * 10.0) / 10.0
      val sign = if value < 0 then "-" else ""
      val whole = rounded.toLong
      val tenth = math.round((rounded - whole) * 10.0)

      s"$sign${grouped(whole)}.$tenth"
    }

  /** Where `value` sits between zero and `max`, for a magnitude bar.
    *
    * A `max` of zero — every row has caught up — gives zero rather than a division by infinity, so a screen
    * where nothing is behind draws no bars at all instead of drawing every bar full.
    */
  def fraction(value: Long, max: Long): Double =
    if max <= 0L then 0.0
    else math.max(0.0, math.min(1.0, value.toDouble / max.toDouble))
}

File: frontend/packages/kernel/src/numbers.ts
/**
 * How KUI prints a quantity.
 *
 * SPEC §6.6 is one sentence long — "numbers are formatted, not narrated" — and it decides more of
 * how a table reads than any colour does. Thousands separators everywhere, so `4212` and `42120`
 * are told apart by their shape rather than by counting digits; a signed figure where the sign is
 * the information; and an em dash, never a zero, where there is no value at all.
 *
 * This lives in the kernel beside `formatBytes` rather than in a feature, because the consumer
 * list, the broker list and the dashboard all print the same kinds of number and three copies of
 * `toLocaleString` is three chances for one screen to group its thousands differently from the one
 * beside it.
 *
 * ## The em dash is a constant and not a literal
 *
 * `MISSING` is exported so that a test can assert on it and so that nobody types a hyphen where the
 * design draws an em dash. "No value" and "a value of nothing" are different facts about a cluster
 * — a consumer group whose lag could not be computed is not a group that has caught up — and the
 * whole reason this string exists is to keep them apart on screen.
 */

/** What a cell holds when there is no value to hold. An em dash, U+2014, not a hyphen. */
export const MISSING = "—";

/**
 * A whole number with thousands separators: `4212` becomes `4,212`.
 *
 * The locale is pinned to `en-US` rather than taken from the browser. KUI prints Kafka's own
 * numbers — offsets, partition ids, lag — beside identifiers that are always ASCII, and a locale
 * that groups with spaces or swaps the decimal separator makes an offset ambiguous when it is
 * pasted into a shell. The prices of that choice is that a French operator sees `4,212`; the price
 * of the other is a support ticket about an offset that would not seek.
 */
export function formatCount(value: number): string {
  if (!Number.isFinite(value)) return MISSING;
  return value.toLocaleString("en-US", { maximumFractionDigits: 0 });
}

/**
 * A rate, to one decimal place, with its sign kept.
 *
 * A negative rate is committed offsets moving *backwards*, which is what somebody else's offset
 * reset looks like from this screen. It is shown as it is rather than clamped to zero, because
 * noticing it is most of the value of printing it at all.
 */
export function formatRate(value: number): string {
  if (!Number.isFinite(value)) return MISSING;
  const rounded = Math.abs(value) < 10 ? Math.round(value * 10) / 10 : Math.round(value);
  return rounded.toLocaleString("en-US", { maximumFractionDigits: 1 });
}

/**
 * A change, with the sign always printed — `+4,212` and `-4,212`.
 *
 * "This rewinds 4 212 records" and "this skips 4 212 records" are the two things an operator is
 * deciding between in the offset-reset preview, and the sign is the entire difference between
 * them. A `+` that is dropped because the number is positive makes the two look the same.
 */
export function formatDelta(value: number): string {
  if (!Number.isFinite(value)) return MISSING;
  if (value === 0) return "0";
  return `${value > 0 ? "+" : "-"}${formatCount(Math.abs(value))}`;
}

/**
 * A fraction of a whole, guarded against the denominator that is not there.
 *
 * `0 / 0` is `NaN`, and `NaN` fails every comparison, so it slips through `Math.min`/`Math.max`
 * unchanged and reaches the stylesheet as a width. A zero denominator therefore yields zero here,
 * which is the reading that cannot mislead: an empty bar for an unknown share, never a full one.
 */
export function share(value: number, of: number): number {
  if (!Number.isFinite(value) || !Number.isFinite(of) || of <= 0) return 0;
  return Math.min(1, Math.max(0, value / of));
}
